### ISSUE TITLE
search: tokens after content is incorrectly parsed

### ISSUE THREAD

**Issue by keegancsmith** [sourcegraph/sourcegraph#9147]
I can't reproduce on sourcegraph.com, but can on sourcegraph.sgdev.org. Very bad regression, since things like clicking "show more" will break the query. I'm not sure when we regressed on this behaviour, but likely related to changes in our parsing code by @rvantonder.

- [`-f:Ingress apiVersion /v1beta`](https://sourcegraph.sgdev.org/search?q=-f:Ingress+apiVersion+/v1beta&patternType=regexp) has results.
- [`apiVersion /v1beta -f:Ingress`](https://sourcegraph.sgdev.org/search?q=apiVersion+/v1beta+-f:Ingress&patternType=regexp) has no results.
- [`f:(DaemonSet|Deployment|ReplicaSet|StatefulSet|NetworkPolicy) apiVersion /v1beta`](https://sourcegraph.sgdev.org/search?q=f:%28DaemonSet%7CDeployment%7CReplicaSet%7CStatefulSet%7CNetworkPolicy%29+apiVersion+/v1beta&patternType=regexp) has results
- [`f:(DaemonSet|Deployment|ReplicaSet|StatefulSet|NetworkPolicy) apiVersion /v1beta count:1000`](https://sourcegraph.sgdev.org/search?q=f:%28DaemonSet%7CDeployment%7CReplicaSet%7CStatefulSet%7CNetworkPolicy%29+apiVersion+/v1beta+count:1000&patternType=regexp) has no results

**Comment by keegancsmith** [comment]
Our tracing isn't ideal here, since we don't get to see the parsed query. But from what we pass on to zoekt you can see that the ` -f:Ingress` is being included as part of the regex.

```
(and (reposet size=169) (and regex:"(apiVersion)(?-s:.)*?(v1beta)" (not file_substr:"Ingress")))

(and (reposet size=169) regex:"(apiVersion)(?-s:.)*?(v1beta -f:Ingress)")
```

Also the pattern type trace logs show the same thing
```
&{Pattern:(apiVersion).*?(v1beta) IsRegExp:true IsStructuralPat:false CombyRule: IsWordMatch:false IsCaseSensitive:false FileMatchLimit:30 IncludePatterns:[] ExcludePattern:Ingress FilePatternsReposMustInclude:[] FilePatternsReposMustExclude:[] PathPatternsAreRegExps:true PathPatternsAreCaseSensitive:false PatternMatchesContent:true PatternMatchesPath:true Languages:[]}

query: &{Pattern:(apiVersion).*?(v1beta -f:Ingress) IsRegExp:true IsStructuralPat:false CombyRule: IsWordMatch:false IsCaseSensitive:false FileMatchLimit:30 IncludePatterns:[] ExcludePattern: FilePatternsReposMustInclude:[] FilePatternsReposMustExclude:[] PathPatternsAreRegExps:true PathPatternsAreCaseSensitive:false PatternMatchesContent:true PatternMatchesPath:true Languages:[]}
```

**Comment by mrnugget** [comment]
Just a note: this only seem to be an issue with regex.

**Comment by mrnugget** [comment]
This bug already exists in `v3.13.2`.

**Comment by mrnugget** [comment]
And it already exists in `v3.12.6`.

**Comment by mrnugget** [comment]
Pretty sure I've nailed it down. We need to escape the `/`. Everything after the `/` is interpreted as part of the regex.:

- [`-f:Ingress apiVersion /v1beta`](https://sourcegraph.sgdev.org/search?q=-f:Ingress+apiVersion+/v1beta&patternType=regexp) has results.
- [`apiVersion /v1beta -f:Ingress`](https://sourcegraph.sgdev.org/search?q=apiVersion+/v1beta+-f:Ingress&patternType=regexp) has no results.
- [`apiVersion \/v1beta -f:Ingress`](https://sourcegraph.sgdev.org/search?q=apiVersion+\/v1beta+-f:Ingress&patternType=regexp) **has results**.
- [`f:(DaemonSet|Deployment|ReplicaSet|StatefulSet|NetworkPolicy) apiVersion /v1beta`](https://sourcegraph.sgdev.org/search?q=f:%28DaemonSet%7CDeployment%7CReplicaSet%7CStatefulSet%7CNetworkPolicy%29+apiVersion+/v1beta&patternType=regexp) has results
- [`f:(DaemonSet|Deployment|ReplicaSet|StatefulSet|NetworkPolicy) apiVersion /v1beta count:1000`](https://sourcegraph.sgdev.org/search?q=f:%28DaemonSet%7CDeployment%7CReplicaSet%7CStatefulSet%7CNetworkPolicy%29+apiVersion+/v1beta+count:1000&patternType=regexp) has no results
- [`f:(DaemonSet|Deployment|ReplicaSet|StatefulSet|NetworkPolicy) apiVersion \/v1beta count:1000`](https://sourcegraph.sgdev.org/search?q=f:%28DaemonSet%7CDeployment%7CReplicaSet%7CStatefulSet%7CNetworkPolicy%29+apiVersion+\/v1beta+count:1000&patternType=regexp) **has results**

I'm not sure whether this is a bug/regression or desired behavior.

**Comment by mrnugget** [comment]
Here's the relevant part of the code where it switches in pattern-scanning-mode: https://github.com/sourcegraph/sourcegraph/blob/8ac72992739396ff92403bea2458d2c84a468ce3/internal/search/query/syntax/scanner.go#L118-L120

**Comment by rvantonder** [comment]
What's happening is that we treat `/.../` syntax specially, see https://docs.sourcegraph.com/user/search/queries#regexp-search. The bug is that the scanner doesn't raise a problem if the `/.../` is 'unterminated'. This behavior has always been this way (at least, since `/.../` was introduced, which is an old addition). The existence of that syntax has caused issues before when I tried to use the scanner for literal search. Looks like it's under-tested and not a regression (or due to changes that I made that I'm aware of).

**Comment by keegancsmith** [comment]
I didn't even know we supported this syntax?

**Comment by rvantonder** [comment]
And yet, it exists for some time! :-) Some history:
- I called it out in [RFC 75](https://docs.google.com/document/d/1iAWIOalYTZLRWmYw0r6_r_uThutLhanyTrxFSFtQAgE/edit#bookmark=id.v7qi1w47eu4f)
- It was an unexpected discovery related to https://github.com/sourcegraph/sourcegraph/issues/6314
- I decided to document it in our docs, instead of removing it

So the current tests is under specified for this syntax, and we don't do enough validation. If we want to keep support (despite the origin being lost to time, I think it's useful?) then we need to check that it is terminated with a `/` like we do with quotes. It also means that anything that starts with `/` needs to be escaped if meant literally (or enclosed in quotes), in regex mode.
